src/server/tasks/queryLLM.task.ts

```diff
@@ -0,0 +1,142 @@
+import { prisma } from "~/server/db";
+import defineTask from "./defineTask";
+import { type CompletionResponse, getCompletion } from "../utils/getCompletion";
+import { type VariableMap, fillTemplateJson } from "../utils/fillTemplate";
+import { type JSONSerializable } from "../types";
+import { sleep } from "../utils/sleep";
+import { shouldStream } from "../utils/shouldStream";
+import { generateChannel } from "~/utils/generateChannel";
+import { reevaluateVariant } from "../utils/evaluations";
+
+const MAX_AUTO_RETRIES = 10;
+const MIN_DELAY = 500; // milliseconds
+const MAX_DELAY = 15000; // milliseconds
+
+function calculateDelay(numPreviousTries: number): number {
+  const baseDelay = Math.min(MAX_DELAY, MIN_DELAY * Math.pow(2, numPreviousTries));
+  const jitter = Math.random() * baseDelay;
+  return baseDelay + jitter;
+}
+
+const getCompletionWithRetries = async (
+  cellId: string,
+  payload: JSONSerializable,
+  channel?: string,
+): Promise<CompletionResponse> => {
+  for (let i = 0; i < MAX_AUTO_RETRIES; i++) {
+    const modelResponse = await getCompletion(payload, channel);
+    if (modelResponse.statusCode !== 429 || i === MAX_AUTO_RETRIES - 1) {
+      return modelResponse;
+    }
+    const delay = calculateDelay(i);
+    await prisma.scenarioVariantCell.update({
+      where: { id: cellId },
+      data: {
+        errorMessage: "Rate limit exceeded",
+        statusCode: 429,
+        retryTime: new Date(Date.now() + delay),
+      },
+    });
+    await sleep(delay);
```

Comment:
It probably doesn't matter for now but we should probably switch to requeuing the job in the future with the [`runAt` parameter](https://github.com/graphile/worker#addjob) set instead of sleeping within a job. That way if there are other jobs waiting to run they can run while we're waiting on this one.

Reply:
Okay, I'll add a TODO